### PR TITLE
getNumberIdSolvedMD

### PR DESCRIPTION
**Pull request by KeruMx**
handler.withDeviceProtocol() returned number of connections if deviceList ==0 is not a registered whatsapp user

### CODE REVIEW

**Comment by KeruMx** [comment]
For #889,

**Comment by mihazavr** [comment]
> For #889,

not works for me. looks like it returns nothing at all

**Comment by jaovitubr** [comment]
Working!

**Comment by KeruMx** [comment]
> > For #889,
> 
> not works for me. looks like it returns nothing at all

what is your error?

**Comment by fm1randa** [comment]
> > > For #889,
> > 
> > 
> > not works for me. looks like it returns nothing at all
> 
> what is your error?

here it's loading infinitely

**Comment by KeruMx** [comment]
> > what is your error?
> 
> here it's loading infinitely

can you put your code, to try replicate

File: src/Client.js [hunk 1]
```diff
@@ -904,10 +904,11 @@ class Client extends EventEmitter {
         }
         return await this.pupPage.evaluate(async (id) => {
             if(window.Store.Features.features.MD_BACKEND){
+                id = window.Store.WidFactory.createWid(id);
                 let handler = (new window.Store.USyncQuery).withContactProtocol();
-                handler = handler.withUser( (new window.Store.USyncUser).withPhone(id),  handler.withBusinessProtocol(), 1 );
+                handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
                 let result = await handler.execute();
-                return result.list[0].contact.type == 'in';
+                return result.list[0].devices.deviceList.length>0;
```
Comment (PurpShell) [inline src/Client.js:911]:
```suggestion
                return result.list[0].devices.deviceList.length > 0;
```
clean the code

File: src/util/Injected.js [hunk 2]
```diff
@@ -74,11 +74,12 @@ exports.LoadUtils = () => {
...
         if (window.Store.Features.features.MD_BACKEND) {
+            id = window.Store.WidFactory.createWid(id);
             let handler = (new window.Store.USyncQuery).withContactProtocol();
-            handler = handler.withUser((new window.Store.USyncUser).withPhone(id), handler.withBusinessProtocol(), 1);
+            handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
             let result = await handler.execute();
-            if (result.list[0].contact.type == 'in') {
-                return result.list[0].id;
+            if (result.list[0].devices.deviceList.length>0) {
+                return id
```
Comment (PurpShell) [inline src/util/Injected.js:82]:
```suggestion
                return id;
```
 missing semicolon

**Review by purpshell — CHANGES_REQUESTED** [review]
Please apply the following changes

File: src/util/Injected.js [hunk 3]
```diff
@@ -74,11 +74,12 @@ exports.LoadUtils = () => {
...
 
         if (window.Store.Features.features.MD_BACKEND) {
+            id = window.Store.WidFactory.createWid(id);
             let handler = (new window.Store.USyncQuery).withContactProtocol();
-            handler = handler.withUser((new window.Store.USyncUser).withPhone(id), handler.withBusinessProtocol(), 1);
+            handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
             let result = await handler.execute();
-            if (result.list[0].contact.type == 'in') {
-                return result.list[0].id;
+            if (result.list[0].devices.deviceList.length>0) {
```
Comment (PurpShell) [inline src/util/Injected.js:81]:
```suggestion
            if (result.list[0].devices.deviceList.length > 0) {
```

**Review by purpshell — APPROVED** [review]
lgtm

**Comment by fm1randa** [comment]
> can you put your code, to try replicate

```ts
import { Client } from "whatsapp-web.js";

async function getChatID(client: Client, number: string) {
	try {
		const numberId = await client.getNumberId(number); // app get stuck here
		return numberId._serialized;
	} catch (error) {
		throw new Error(
			`Couldn't find chatID for number ${number}. Make sure the number is registered in Whatsapp or the number is in correct format (+5521987654321).\n\n${error}`
		);
	}
}

export default getChatID;

```

**Comment by KeruMx** [comment]
> > can you put your code, to try replicate
> 
> ```ts
> import { Client } from "whatsapp-web.js";
> 
> async function getChatID(client: Client, number: string) {
> 	try {
> 		const numberId = await client.getNumberId(number); // app gets stuck here
> 		return numberId._serialized;
> 	} catch (error) {
> 		throw new Error(
> 			`Couldn't find chatID for number ${number}. Make sure the number is registered in Whatsapp or the number is in correct format (+5521987654321).\n\n${error}`
> 		);
> 	}
> }
> 
> export default getChatID;
> ```

Only works for MultiDevice Branch, are you using this branch?

**Comment by fm1randa** [comment]
> Only works for MultiDevice Branch, are you using this branch?
I was using your branch `KeruMx:getNumberIdMD` and after the merge, at `pedroslopez:multidevice` got the same result :(

**Comment by ReniDelonzek** [comment]
I tried this and it didn't work for me either.

**Comment by Shvan11** [comment]
getNumberID() is not working for me too, it is stuck at this line:

`` handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
            let result = await handler.execute();``

**Comment by fm1randa** [comment]
> getNumberID() is not working for me too, it is stuck at this line:
> 
> handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
> 
> ```
>         let result = await handler.execute();`
> ```

same here.

**Comment by mihazavr** [comment]
Any ideas how to check if number registered another way?

**Comment by alissonbezerra** [comment]
> > getNumberID() is not working for me too, it is stuck at this line:
> > handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
> > ```
> >         let result = await handler.execute();`
> > ```
> 
> same here.

Seems like the problem is in `Client.js`. 

```
        if(!number.startsWith('+')) {
            number = '+' + number;
        }
```

When I removed those lines above from the function `getNumberId`, the Promise was returned successfully.

**Comment by mihazavr** [comment]
> > > getNumberID() is not working for me too, it is stuck at this line:
> > > handler = handler.withUser((new window.Store.USyncUser).withId(id), handler.withDeviceProtocol(), 1);
> > > ```
> > >         let result = await handler.execute();`
> > > ```
> > 
> > 
> > same here.
> 
> Seems like the problem is in `Client.js`.
> 
> ```
>         if(!number.startsWith('+')) {
>             number = '+' + number;
>         }
> ```
> 
> When I removed those lines above from the function `getNumberId`, the Promise was returned successfully.

It works! Thank you!

**Comment by ReniDelonzek** [comment]
Hello guys, I submitted a [PR](https://github.com/pedroslopez/whatsapp-web.js/pull/1159) applying the fix, thank you @alissonbezerra  for pointing out the problem

**Comment by mahmouds12** [comment]
@ReniDelonzek 
client.getNumberId() is still not working as expected. It works correctly for **correct whatsapp numbers**, but if the number is incorrect, it will stuck forever

**Comment by PurpShell** [comment]
> @ReniDelonzek 
> client.getNumberId() is still not working as expected. It works correctly for **correct whatsapp numbers**, but if the number is incorrect, it will stuck forever

Then we can place a timeout

**Pull request merged**